src/Featurizers/ShortGrainDropperFeaturizer.h

```diff
@@ -0,0 +1,216 @@
+
+    void complete_training_impl(void) override;
+
+    // MSVC has problems when the definition is separate from the declaration
+    typename BaseType::TransformerUniquePtr create_transformer_impl(void) override {
+        return typename BaseType::TransformerUniquePtr(new TransformerType(std::move(_grainsToDrop)));
+    }
+};
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// --------------------------------------------------------------------
+// |
+// |  ShortGrainDropperEstimator
+// |
+// ----------------------------------------------------------------------
+template <size_t MaxNumTrainingItemsV>
+ShortGrainDropperEstimator<MaxNumTrainingItemsV>::ShortGrainDropperEstimator(
+    AnnotationMapsPtr pAllColumnAnnotations,
+    size_t colIndex,
+    std::uint8_t windowSize,
+    std::vector<std::uint8_t> lags,
+    std::uint8_t maxHorizon,
+    nonstd::optional<std::uint8_t> cv
+) :
+    BaseType("ShortGrainDropperEstimatorImpl", std::move(pAllColumnAnnotations)),
+    _colIndex(
+        [this, &colIndex](void) -> size_t & {
+            if(colIndex >= this->get_column_annotations().size())
+                throw std::invalid_argument("colIndex");
+            return colIndex;
+        }()
+    ),
+    _minPoints(
+        [&windowSize, &lags, &maxHorizon, &cv](void) -> std::uint16_t {
+            //it appears automl tests show that
+            //windowSize can be 0
+            //lags could contain 0s
+            //maxHorizon may not be 0, not sure currently
+            //cv may not be 0, not sure currently
+            if (lags.size() == 0)
+                throw std::invalid_argument("lags");
+            if (!cv.has_value())
+                return (maxHorizon + std::max(windowSize, *std::max_element(lags.cbegin(), lags.cend())) + 1);
+            return (2*maxHorizon + static_cast<std::uint8_t>(*cv) + std::max(windowSize, *std::max_element(lags.cbegin(), lags.cend())) + 1);
+        }()
+    ) {
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <size_t MaxNumTrainingItemsV>
+bool ShortGrainDropperEstimator<MaxNumTrainingItemsV>::begin_training_impl(void) /*override*/ {
+    return true;
+}
+
+template <size_t MaxNumTrainingItemsV>
+void ShortGrainDropperEstimator<MaxNumTrainingItemsV>::complete_training_impl(void) /*override*/ {
+    for (GrainsMap::value_type const & groupByGrainsElement : _groupByGrains) {
```

Comment:
>_groupByGrains [](start = 62, length = 14)

We don't need to keep _groupByGrains once _grainsToDrop is complete; consider clearing _groupByGrains at the end of this loop.